test/state/mpt.cpp

```diff
@@ -0,0 +1,279 @@
+// evmone: Fast Ethereum Virtual Machine implementation
+// Copyright 2022 The evmone Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "mpt.hpp"
+#include "rlp.hpp"
+#include <algorithm>
+#include <cassert>
+
+namespace evmone::state
+{
+namespace
+{
+/// The collection of nibbles (4-bit values) representing a path in a MPT.
+struct Path
+{
+    size_t length = 0;  // TODO: Can be converted to uint8_t.
+    uint8_t nibbles[64]{};
+
+    Path() = default;
+
+    explicit Path(bytes_view key) noexcept : length{2 * key.size()}
+    {
+        assert(length <= std::size(nibbles));
+        size_t i = 0;
+        for (const auto b : key)
+        {
+            // static_cast is only needed in GCC <= 8.
+            nibbles[i++] = static_cast<uint8_t>(b >> 4);
+            nibbles[i++] = static_cast<uint8_t>(b & 0x0f);
+        }
+    }
+
+    [[nodiscard]] Path tail(size_t pos) const noexcept
+    {
+        assert(pos <= length);
+        Path p;
+        p.length = length - pos;
+        std::copy_n(&nibbles[pos], p.length, p.nibbles);
+        return p;
+    }
+
+    [[nodiscard]] Path head(size_t size) const noexcept
+    {
+        assert(size < length);
```

Comment:
The assert comes from how this is used in MPT: using `head()` or `tail()` it may produce empty path but never produces a full copy of a path. I added separate asserts for this scope narrowing.